utils/writeable/src/write_or_ref.rs

```diff
@@ -0,0 +1,203 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::Writeable;
+use alloc::borrow::Cow;
+use alloc::string::String;
+use core::fmt;
+
+/// Bytes that have been partially validated as UTF-8 up to an offset.
+struct PartiallyValidatedUtf8<'a> {
+    // Safety Invariants:
+    // 1. The offset is less than or equal to the length of the slice.
+    // 2. The slice is valid UTF-8 up to the offset.
+    slice: &'a [u8],
+    offset: usize,
+}
+
+impl<'a> PartiallyValidatedUtf8<'a> {
+    fn new(slice: &'a [u8]) -> Self {
+        Self { slice, offset: 0 }
+    }
+
+    /// Check whether the given string is the next chunk of unvalidated bytes.
+    /// If so, increment offset and return true. Otherwise, return false.
+    fn try_push(&mut self, valid_str: &str) -> bool {
+        if self.slice.get(self.offset..self.offset + valid_str.len()) == Some(valid_str.as_bytes())
+        {
+            // Safety Invariant: `valid_str` is valid UTF-8, and we are
+            // incrementing the offset to cover bytes equal to `valid_str`
+            self.offset += valid_str.len();
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Return the validated portion as `&str`.
+    fn validated_as_str(&self) -> &'a str {
+        debug_assert!(self.offset <= self.slice.len());
+        // Safety: self.offset is a valid end index in a range
+        let valid_slice = unsafe { self.slice.get_unchecked(..self.offset) };
+        debug_assert!(core::str::from_utf8(valid_slice).is_ok());
+        // Safety: the UTF-8 of slice has been validated up to offset
+        unsafe { core::str::from_utf8_unchecked(valid_slice) }
+    }
+}
+
+enum SliceOrString<'a> {
+    Slice(PartiallyValidatedUtf8<'a>),
+    String(String),
+}
+
+/// This is an infallible impl. Functions always return Ok, not Err.
+impl fmt::Write for SliceOrString<'_> {
+    #[inline]
+    fn write_str(&mut self, other: &str) -> fmt::Result {
+        let owned: &mut String = match self {
+            SliceOrString::Slice(slice) => {
+                if slice.try_push(other) {
+                    return Ok(());
+                }
+                // We failed to match. Convert to owned, put it in the field,
+                // and get it out again.
+                let valid_str = slice.validated_as_str();
+                let owned = String::from(valid_str);
+                *self = SliceOrString::String(owned);
+                let SliceOrString::String(owned) = self else {
+                    unreachable!()
+                };
+                owned
+            }
+            SliceOrString::String(owned) => owned,
+        };
+        owned.write_str(other)
```

Comment:
this is jumping through a lot of hoops to not duplicate line 75. I think the code would be a lot easier if you did that:

```suggestion
        match self {
            SliceOrString::Slice(slice) => {
                if !slice.try_push(other) {
                    // We failed to match. Convert to owned.
                    let valid_str = slice.validated_as_str();
                    let owned = format!("{valid_str}{other});
                    *self = SliceOrString::String(owned);
                    Ok(())
            }
            SliceOrString::String(owned) => owned.write_str(other),
        }
```

Reply:
I don't think it's a lot of hoops: it is a standard "base case" coding style. That said, I think your code is more readable, except I don't like using `format!`. I'll push up something similar.

Reply:
For this Rust program:

```rust
#[inline(never)]
fn string_from_pieces_format(a: &str, b: &str) -> String {
    std::format!("{a}{b}")
}

#[inline(never)]
fn string_from_pieces_manual(a: &str, b: &str) -> String {
    let mut s = String::with_capacity(a.len() + b.len());
    s.push_str(a);
    s.push_str(b);
    s
}

fn main() {
    let s1 = string_from_pieces_format("hello", "world");
    let s2 = string_from_pieces_manual("hello", "world");
    println!("{s1} {s2}");
}
```

Here is the ASM (https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=dff4d0ae823abc03d88fad4b84b56a7d):

```asm
std::rt::lang_start:
	pushq	%rax
	movl	%ecx, %r8d
	movq	%rdx, %rcx
	movq	%rsi, %rdx
	movq	%rdi, (%rsp)
	leaq	.L__unnamed_1(%rip), %rsi
	movq	%rsp, %rdi
	callq	*std::rt::lang_start_internal@GOTPCREL(%rip)
	popq	%rcx
	retq

std::rt::lang_start::{{closure}}:
	pushq	%rax
	movq	(%rdi), %rdi
	callq	std::sys::backtrace::__rust_begin_short_backtrace
	xorl	%eax, %eax
	popq	%rcx
	retq

std::sys::backtrace::__rust_begin_short_backtrace:
	pushq	%rax
	callq	*%rdi
	#APP
	#NO_APP
	popq	%rax
	retq

<&T as core::fmt::Display>::fmt:
	movq	%rsi, %rdx
	movq	(%rdi), %rax
	movq	8(%rdi), %rsi
	movq	%rax, %rdi
	jmpq	*<str as core::fmt::Display>::fmt@GOTPCREL(%rip)

core::ops::function::FnOnce::call_once{{vtable.shim}}:
	pushq	%rax
	movq	(%rdi), %rdi
	callq	std::sys::backtrace::__rust_begin_short_backtrace
	xorl	%eax, %eax
	popq	%rcx
	retq

<alloc::string::String as core::fmt::Display>::fmt:
	movq	%rsi, %rdx
	movq	8(%rdi), %rax
	movq	16(%rdi), %rsi
	movq	%rax, %rdi
	jmpq	*<str as core::fmt::Display>::fmt@GOTPCREL(%rip)

playground::string_from_pieces_format:
	subq	$120, %rsp
	leaq	.L__unnamed_2(%rip), %rax
	movq	%rax, 8(%rsp)
	movq	$5, 16(%rsp)
	leaq	.L__unnamed_3(%rip), %rax
	movq	%rax, 24(%rsp)
	movq	$5, 32(%rsp)
	leaq	8(%rsp), %rax
	movq	%rax, 40(%rsp)
	leaq	<&T as core::fmt::Display>::fmt(%rip), %rax
	movq	%rax, 48(%rsp)
	leaq	24(%rsp), %rcx
	movq	%rcx, 56(%rsp)
	movq	%rax, 64(%rsp)
	leaq	.L__unnamed_4(%rip), %rax
	movq	%rax, 72(%rsp)
	movq	$2, 80(%rsp)
	movq	$0, 104(%rsp)
	leaq	40(%rsp), %rax
	movq	%rax, 88(%rsp)
	movq	$2, 96(%rsp)
	leaq	72(%rsp), %rsi
	callq	*alloc::fmt::format::format_inner@GOTPCREL(%rip)
	addq	$120, %rsp
	retq

playground::string_from_pieces_manual:
	pushq	%rbx
	movq	%rdi, %rbx
	movq	__rust_no_alloc_shim_is_unstable@GOTPCREL(%rip), %rax
	movzbl	(%rax), %eax
	movl	$10, %edi
	movl	$1, %esi
	callq	*__rust_alloc@GOTPCREL(%rip)
	testq	%rax, %rax
	je	.LBB7_2
	movb	$111, 4(%rax)
	movl	$1819043176, (%rax)
	movl	$1819438967, 5(%rax)
	movb	$100, 9(%rax)
	movq	$10, (%rbx)
	movq	%rax, 8(%rbx)
	movq	$10, 16(%rbx)
	popq	%rbx
	retq

.LBB7_2:
	movl	$1, %edi
	movl	$10, %esi
	callq	*alloc::raw_vec::handle_error@GOTPCREL(%rip)

playground::main:
	pushq	%r14
	pushq	%rbx
	subq	$136, %rsp
	leaq	8(%rsp), %rbx
	movq	%rbx, %rdi
	callq	playground::string_from_pieces_format
	leaq	32(%rsp), %r14
	movq	%r14, %rdi
	callq	playground::string_from_pieces_manual
	movq	%rbx, 56(%rsp)
	leaq	<alloc::string::String as core::fmt::Display>::fmt(%rip), %rax
	movq	%rax, 64(%rsp)
	movq	%r14, 72(%rsp)
	movq	%rax, 80(%rsp)
	leaq	.L__unnamed_5(%rip), %rax
	movq	%rax, 88(%rsp)
	movq	$3, 96(%rsp)
	movq	$0, 120(%rsp)
	leaq	56(%rsp), %rax
	movq	%rax, 104(%rsp)
	movq	$2, 112(%rsp)
	leaq	88(%rsp), %rdi
	callq	*std::io::stdio::_print@GOTPCREL(%rip)
	movq	32(%rsp), %rsi
	testq	%rsi, %rsi
	je	.LBB8_4
	movq	40(%rsp), %rdi
	movl	$1, %edx
	callq	*__rust_dealloc@GOTPCREL(%rip)

.LBB8_4:
	movq	8(%rsp), %rsi
	testq	%rsi, %rsi
	je	.LBB8_6
	movq	16(%rsp), %rdi
	movl	$1, %edx
	callq	*__rust_dealloc@GOTPCREL(%rip)

.LBB8_6:
	addq	$136, %rsp
	popq	%rbx
	popq	%r14
	retq
	movq	%rax, %rbx
	movq	32(%rsp), %rsi
	testq	%rsi, %rsi
	je	.LBB8_8
	movq	40(%rsp), %rdi
	movl	$1, %edx
	callq	*__rust_dealloc@GOTPCREL(%rip)
	jmp	.LBB8_8
	movq	%rax, %rbx

.LBB8_8:
	movq	8(%rsp), %rsi
	testq	%rsi, %rsi
	je	.LBB8_10
	movq	16(%rsp), %rdi
	movl	$1, %edx
	callq	*__rust_dealloc@GOTPCREL(%rip)

.LBB8_10:
	movq	%rbx, %rdi
	callq	_Unwind_Resume@PLT

main:
	pushq	%rax
	movq	%rsi, %rcx
	movslq	%edi, %rdx
	leaq	playground::main(%rip), %rax
	movq	%rax, (%rsp)
	leaq	.L__unnamed_1(%rip), %rsi
	movq	%rsp, %rdi
	xorl	%r8d, %r8d
	callq	*std::rt::lang_start_internal@GOTPCREL(%rip)
	popq	%rcx
	retq

.L__unnamed_1:
	.asciz	"\000\000\000\000\000\000\000\000\b\000\000\000\000\000\000\000\b\000\000\000\000\000\000"
	.quad	core::ops::function::FnOnce::call_once{{vtable.shim}}
	.quad	std::rt::lang_start::{{closure}}
	.quad	std::rt::lang_start::{{closure}}

.L__unnamed_4:
	.quad	1
	.zero	8
	.quad	1
	.zero	8

.L__unnamed_2:
	.ascii	"hello"

.L__unnamed_3:
	.ascii	"world"

.L__unnamed_6:
	.byte	32

.L__unnamed_7:
	.byte	10

.L__unnamed_5:
	.quad	1
	.zero	8
	.quad	.L__unnamed_6
	.asciz	"\001\000\000\000\000\000\000"
	.quad	.L__unnamed_7
	.asciz	"\001\000\000\000\000\000\000"
```

So it appears avoiding `format!` is worthwhile.

Reply:
Cannot parse that atm. Just make sure to reserve capacity once.